File: solutions/c/rotational-cipher/1/rotational_cipher.c
#include "rotational_cipher.h"
#include <string.h>
#include <ctype.h>
#include <stdlib.h>
char *rotate(const char *text, int shift_key){
    int len = strlen(text);
    char *result = calloc(len+1, sizeof(char));
    strcpy(result, text);
    for (int i = 0; i < len; i++) {
        if (isalpha(text[i])) {
            char base = islower(text[i]) ? 'a' : 'A';
            result[i] = (text[i] - base + shift_key) % 26;
            if (result[i] < 0) result[i] += 26;
            result[i] += base;
        } else {
            result[i] = text[i];
        }
    }
    return result;
}

File: solutions/zig/rotational-cipher/1/rotational_cipher.zig
const std = @import("std");
const mem = std.mem;

pub fn rotate(allocator: mem.Allocator, text: []const u8, shiftKey: u5) mem.Allocator.Error![]u8 {
    const len = text.len;
    const result = try allocator.alloc(u8, len);
    errdefer allocator.free(result);

    for (text, result) |t, *r| {
        if (std.ascii.isAlphabetic(t)) {
            const base: u8 = if (std.ascii.isUpper(t)) 'A' else 'a';
            r.* = base + @as(u8, @intCast((t - base + shiftKey) % 26));
        } else r.* = t;
    }

    return result;
}
